File: rubycalc/lib/ast.rb
require 'set'
require 'calculator'

class BinaryNode
  attr_reader :left, :right
  
  def initialize(left,right)
    @left = left
    @right = right
  end
end
   
class UnaryNode
  attr_reader :subTree
   
  def initialize subTree
    @subTree = subTree
  end
end

class AddNode < BinaryNode
  def initialize(left, right)
    super left,right
  end
   
  def evaluate
    @left.evaluate + @right.evaluate
  end
end

class SubNode < BinaryNode
  def initialize left, right
    super left,right
  end
   
  def evaluate 
    @left.evaluate - @right.evaluate
  end
end

class TimesNode < BinaryNode
  def initialize left, right
    super left,right
  end
   
  def evaluate 
    @left.evaluate * @right.evaluate
  end
end

class DivideNode < BinaryNode
  def initialize left, right
    super left,right
  end
   
  def evaluate
    if @right.evaluate == 0.0 then
      raise Inf
    else
      @left.evaluate / @right.evaluate
    end
  end
end

class ModuleNode < BinaryNode
  def initialize left, right
    super left,right
  end

  def evaluate
    @left.evaluate.modulo(@right.evaluate)
  end
end

class NumNode 
  def initialize(num)
    @num = num
  end
   
  def evaluate
    @num
  end
end

class EolNode
  def initialize(eol)
    @eol = eol
  end

  def evaluate
    @eol
  end
end

class IDNode
  attr_accessor :ro
  def initialize(ro)
    @ro = ro
  end

  def evaluate
    pt = $calc.potatos[ro]
    pt
  end
end

class NonNode < UnaryNode
  def initialize sub
    super sub
  end

  def evaluate
    @subTree.evaluate * -1.0
  end
end

class StoreNode < UnaryNode
  def initialize sub
    super sub
  end

  def evaluate
    $calc.memory = @subTree.evaluate
    $calc.memory
  end
end

class PlusNode < UnaryNode
  def initialize sub
    super sub
  end

  def evaluate
    $calc.memory = $calc.memory + @subTree.evaluate
    $calc.memory
  end
end

class MinusNode < UnaryNode
  def initialize sub
    super sub
  end

  def evaluate
    $calc.memory = $calc.memory - @subTree.evaluate
    $calc.memory
  end
end

class RecallNode
  def initialize; end

  def evaluate
    $calc.memory
  end
end

class ClearNode
  def initialize; end

  def evaluate
    $calc.memory = 0.0
  end
end

class AssignationNode < UnaryNode
  attr_accessor :pato
  def initialize(sub, pato)
    super sub
    
    @pato = pato
  end

  def evaluate
    co = @subTree.evaluate
    $calc.potatos.store(@pato, co)
    return co
  end
end

class ABSNode < UnaryNode
  def initialize sub
    super sub
  end

  def evaluate
    @subTree.evaluate.abs
  end
end

class SINNode < UnaryNode
  def initialize sub
    super sub
  end

  def evaluate
    Math.sin(@subTree.evaluate)
  end
end

class COSNode < UnaryNode
  def initialize sub
    super sub
  end

  def evaluate
    Math.cos(@subTree.evaluate)
  end
end

class TANNode < UnaryNode
  def initialize sub
    super sub
  end

  def evaluate
    Math.tan(@subTree.evaluate)
  end
end

class EXPNode < UnaryNode
  def initialize sub
    super sub
  end

  def evaluate
    Math.exp(@subTree.evaluate)
  end
end

class LOGNode < UnaryNode
  def initialize sub
    super sub
  end

  def evaluate
    Math.log(@subTree.evaluate)
  end
end

class SQRTNode < UnaryNode
  def initialize sub
    super sub
  end

  def evaluate
    if @subTree.evaluate < 0.0 then
      raise Nan
    else
      Math.sqrt(@subTree.evaluate)
    end
  end
end
    

File: cppcalc/ast.cpp
#include "ast.h"
#include <iostream>
#include <math.h> 
#include "calculator.h"

// for debug information uncomment
//#define debug

AST::AST() {}

AST::~AST() {}

BinaryNode::BinaryNode(AST* left, AST* right):
   AST(),
   leftTree(left),
   rightTree(right)
{}

BinaryNode::~BinaryNode() {
#ifdef debug
   cout << "In BinaryNode destructor" << endl;
#endif

   try {
      delete leftTree;
   } catch (...) {}

   try {
      delete rightTree;
   } catch(...) {}
}
   
AST* BinaryNode::getLeftSubTree() const {
   return leftTree;
}

AST* BinaryNode::getRightSubTree() const {
   return rightTree;
}

UnaryNode::UnaryNode(AST* sub):
   AST(),
   subTree(sub)
{}

UnaryNode::~UnaryNode() {
#ifdef debug
   cout << "In UnaryNode destructor" << endl;
#endif

   try {
      delete subTree;
   } catch (...) {}
}
   
AST* UnaryNode::getSubTree() const {
   return subTree;
}

AddNode::AddNode(AST* left, AST* right):
   BinaryNode(left,right)
{}

float  AddNode::evaluate() {
   return getLeftSubTree()->evaluate() + getRightSubTree()->evaluate();
}

SubNode::SubNode(AST* left, AST* right):
   BinaryNode(left,right)
{}

float SubNode::evaluate() {
   return getLeftSubTree()->evaluate() - getRightSubTree()->evaluate();
}

TimesNode::TimesNode(AST* left, AST* right):
  BinaryNode(left,right)
{}

float TimesNode::evaluate() {
   return getLeftSubTree()->evaluate() * getRightSubTree()->evaluate();
}


DivideNode::DivideNode(AST* left, AST* right):
  BinaryNode(left,right)
{}

float DivideNode::evaluate() {
  float n = getRightSubTree()->evaluate();
  if(n == 0){
    cerr << "inf" << endl;
  }
   return getLeftSubTree()->evaluate() / getRightSubTree()->evaluate();
}

ModuleNode::ModuleNode(AST* left, AST* right):
  BinaryNode(left, right)
{}

float ModuleNode::evaluate() {
  //return getLeftSubTree()->evaluate() % getRightSubTree()->evaluate();
  return fmod(getLeftSubTree()->evaluate(), getRightSubTree()->evaluate());
}

ABSNode::ABSNode(AST* ast) :
  UnaryNode(ast) { }

float ABSNode::evaluate() {
   float n = getSubTree()->evaluate();
   n = fabs (n);
  return n;
}


SINNode::SINNode(AST* ast) :
  UnaryNode(ast) { }


float SINNode::evaluate() {
  float n = sin(getSubTree()->evaluate());
  return n;
}

COSNode::COSNode(AST* ast) :
  UnaryNode(ast) { }


float COSNode::evaluate() {
  float n = cos(getSubTree()->evaluate());
  return n;
}

TANNode::TANNode(AST* ast) :
  UnaryNode(ast) { }


float TANNode::evaluate() {
  float PI = 3.14159265;
  float n = tan(getSubTree()->evaluate() * PI / 180.0);
  return n;
}

EXPNode::EXPNode(AST* ast) :
  UnaryNode(ast) { }


float EXPNode::evaluate() {
  float n = exp(getSubTree()->evaluate());
  return n;
}

LOGNode::LOGNode(AST* ast) :
  UnaryNode(ast) { }


float LOGNode::evaluate() {
  float n = log(getSubTree()->evaluate());
  return n;
}

SQRTNode::SQRTNode(AST* ast) :
  UnaryNode(ast) { }


float SQRTNode::evaluate() {
  float n = sqrt(getSubTree()->evaluate());
  if (n < 0) {
    cout << "nan" <<endl;
  }
  return n;
}


StoreNode::StoreNode(AST* ast) :
  UnaryNode(ast) { }

float StoreNode::evaluate() {
  
  calc->store(getSubTree()->evaluate());
  return calc->recall();
}

NonNode::NonNode(AST* ast) :
  UnaryNode(ast) { }

float NonNode::evaluate() {
  return getSubTree()->evaluate() * -1;
  //return getSubTree()->evaluate() * -1.0;
  }

PlusNode::PlusNode(AST* ast) :
  UnaryNode(ast) { }

float PlusNode::evaluate() {
  float x = calc->recall();
  calc->store(x + getSubTree()->evaluate());
  return calc->recall();
}

MinusNode::MinusNode(AST* ast) :
  UnaryNode(ast) { }

float MinusNode::evaluate() {
  float x = calc->recall();
  calc->store(x - getSubTree()->evaluate());
  return calc->recall();
}

ClearNode::ClearNode() : AST() { }

float ClearNode::evaluate() {
  return calc->clear();
} 

NumNode::NumNode(float n) :
   AST(),
   val(n)
{}

float NumNode::evaluate() {
   return val;
}

IDNode::IDNode(string m) :
  AST(),
  val(m)
{}

float IDNode::evaluate() {
  return calc->mapS(val);
}

AssignationNode::AssignationNode(AST* ast, string val):
  UnaryNode(ast),
  val(val)
{}

float AssignationNode::evaluate() {
  float n = getSubTree()->evaluate();
  calc->setValue(val, n);
  return n;
}

RecallNode::RecallNode() : AST() { }

float RecallNode::evaluate() {
  return calc->recall();
}

			 
